content/blog/2019-11-13-space-efficient-gc/index.md

```diff
@@ -0,0 +1,100 @@
+4. According to the paper, blacklist can be easily incorpoated into a garbage collecting allocator at *almost no performance cost*.
+
+### Other Sources of Excess Retention
+
+Another source of excess memory retention is due to the fact that conservative collectors usually have less information about variable liveness than conventional collectors. As it lacks these information, the allocator could come for help.
+
+Here are some useful techniques:
+- Have the allocator and collector carefully clean up after themselves, clearing local variables before function exit.
+- The allocator try to clear areas sometimes in the stack beyond the most recently activated frame.
+
+### Minor: Consequences of Misidentification
+
+Actually, the involved data structure can greatly influence the individual false reference. 
+
+For example, given a balanced binary tree, the expected number of vertices retained in a false reference is about the height of the tree, which is usually tolerable. Queses and lazy lists could exhisit much worse behavior as they grow without bound.
+
+A more common problem is the construction of largely stronly connected data structures, which could result in an unvounded memory leak if the structures are large enough. Take the data structures shown in the below graph as an example.
+
+![004.png](./004.png)
+
+They are a rectangular array of vertices, which are linked both horizontally and vertically. The structure can be accessed by traversing a row/column. The left shows an embedded link representation. A false reference in this structure would result in the retention of a large fraction of the whold structure. The right shows an embedded link representation of the same structure, with a separate link representation (represented by ovals in the fig). Thus, at most a single row/column is affected.
+
+Therefore, the embedded link version of some data structures would greatly help reduce storage retention and should be encouraged.
+
+# Conclusion and Thoughts
+
+This paper introduces some simple but effective techniques for reducing storage retention in conservative garbage collectors. 
+
+- I think this paper gives a thorough explanation in the pointer misidentification problem in conservative garbage collectors, as well as the current and their proposed solutions. The blacklisting solution is simple and effective, and can be easily incorporated into current collectors. 
+- However, it doesn't give a thorough evaluation of their techniques. Only a handwritten program is tested for the effectiveness and no experiment is conducted on the performance cost when incorporating it into current garbage collectors. Also, they don't consider the possible memory fragmentation problem in evaluation.
+
+# Questions
+- How much could fragmentation due to blacklisting influence the performance?
+- How do you like the organization/design of the evaluation? (I thought it quite inadequate.)
```

Comment:
It is, certainly, by 2019 standards. But 1993 was a very different time. To be blunt: I don't think it's worth asking "is this a good evaluation?" because all right-thinking 6120 students will answer "no; there's not a single bar chart and they didn't even use a benchmark suite."